src/main.cpp

```diff
@@ -5708,89 +5677,90 @@ bool GetEarliestStakeTime(std::string grcaddress, std::string cpid)
     return true;
 }
 
-HashSet GetCPIDBlockHashes(const std::string& cpid)
+void AddRARewardBlock(const CBlockIndex* pindex)
 {
-    auto hashes = mvCPIDBlockHashes.find(cpid);
-    return hashes != mvCPIDBlockHashes.end()
-        ? hashes->second
-        : HashSet();
-}
+    // this is from LoadBlockIndex
+    if (pindex->nResearchSubsidy > 0 && pindex->IsUserCPID())
+    {
+        const std::string& cpid = pindex->GetCPID();
 
-void AddCPIDBlockHash(const std::string& cpid, const uint256& blockhash)
-{
-    // Add block hash to CPID hash set.
-    mvCPIDBlockHashes[cpid].emplace(blockhash);
+        StructCPID& stCPID = GetInitializedStructCPID2(cpid,mvResearchAge);
+
+        stCPID.InterestSubsidy += pindex->nInterestSubsidy;
+        stCPID.ResearchSubsidy += pindex->nResearchSubsidy;
+        if (pindex->nHeight > stCPID.LastBlock)
+        {
+            stCPID.LastBlock = pindex->nHeight;
+            stCPID.BlockHash = pindex->GetBlockHash().GetHex();
+        }
+
+        if (pindex->nMagnitude > 0)
+        {
+            stCPID.Accuracy++;
+            stCPID.TotalMagnitude += pindex->nMagnitude;
+        }
+
+        if (pindex->nTime < stCPID.LowLockTime)  stCPID.LowLockTime = pindex->nTime;
+        if (pindex->nTime > stCPID.HighLockTime) stCPID.HighLockTime = pindex->nTime;
+
+        // Add block hash to CPID hash set.
+        stCPID . vRewardBlocs . emplace(pindex);
+        // Store the updated struct. (implicit)
+    }
 }
 
-void RemoveCPIDBlockHash(const std::string& cpid, const uint256& blockhash)
+void RemoveCPIDBlockHash(const std::string& cpid, const CBlockIndex* pindex)
 {
-   mvCPIDBlockHashes[cpid].erase(blockhash);
+    //TODO: is copying involved here?
```

Comment:
Copy is indeed involved here. We could perhaps return a reference so we don't constantly have to Get->Update->Set when handling StructCPID ojects.

Reply:
He is fast! Makes me feel like a slug! :)